### DISCUSSION TITLE
Questions about control flow of ARM Interrupts

### DISCUSSION THREAD

**Discussion by vrindisbacher** [tock/tock#4242] [Q&A] [answered]
Hey there - I have been looking at the ISR implementations for ARMv7m and ARMv6m, and I had some questions about what control flow looks like for interrupts. 

I am specifically interested in what happens when processes are pre-empted by an interrupt. Here is my current understanding (for ARMv6m).   

- Switching to user space involves running [the following assembly](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L380).
- In this assembly, we restore the value of previously saved register r4 - r11 from the `Process` struct’s stored registers field 
- Then, we make an `svc` call which switches to user land using [the following assembly](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L223).
- Then, say an interrupt happens, we will immediately enter the ISR which executes [the following assembly](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L89). This top half handler saves the values of register r4 - r11 into the `Process` struct’s stored registers field, and  marks the interrupt as pending so it can be handled by the bottom half handler. 
- When we finish executing, we resume on line 406 of [the `switch_to_user` assembly](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L406). This assembly once again saves the values of r4 - r11 into the `Process` struct’s stored registers field.

Is this an accurate model of what is happening? If so, I have a few follow up questions. 

1. When we run the ISR, how do we know that `sp + 4 bytes` has the address of the correct field in the `Process` struct? 
   - My intuition here is that when we switch to user land, hardware stacking takes care of saving the current value of `r1`  on the main stack (used by the kernel) and switch to the process stack to run the process. When we call the ISR in interrupt, we switch back to the main stack which still has the value of `r1` at a 4 byte offset. Is this correct? 
3. Why do we have to save register values in [the ARMv6 ISR](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L89)? If we yield back to the `switch_to_user` assembly, aren’t we handling saving those registers there? It seems that [the ARMv7 ISR](https://github.com/PLSysSec/tock/blob/905ae8ced54eab523bc9365df413f8fd2bec4331/arch/cortex-v7m/src/lib.rs#L152) does not store registers which I find confusing… Is there a reason for this difference? 
4. How do we resume on line 406 of [the `switch_to_user` assembly](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L406)? 
    - My intuition here is that we are using exec return value of `0xFFFFFFF9` which instructs ARM to resume with state from the main stack. The main stack’s state is the state at the point we made the supervisor call in the `switch_to_user` assembly. Is this correct? 
    
Please let me know if I can clarify anything. Thanks for taking the time to look at this!!

**Comment by ppannuto** [comment, marked as answer]
First/meta note: the v7m pathway is _much_ more tested than the v6m pathway. ~~While there are no known issues with v6m~~, for a window of time we had no v6m boards, and now the only board that does use it is variants of the RPi Pico—I don't know how regularly or aggressively exercised that is. This is to say, if you're going to focus a deep dive, I'd prioritize v7m.

That said, in answering your questions, I believe you've identified a bug in the v6m implementation, more below:

---

> 1. When we run the ISR, how do we know that `sp + 4` bytes has the address of the correct field in the `Process` struct? 

Yes, the intuition is correct. `MSP`=`Main Stack Pointer` which is the stack used by the kernel (vs `PSP`=`Process Stack Pointer`); these are ARM terminology. See `B1.4.1 The ARM core registers` in the `ARM ARM` (The ARM Architecture Reference Manual) [same section number and heading in ARMv6 ARM and ARMv6 ARM] for details on the dual-stack setup.

> 2. Why do we have to save register values in [the ARMv6 ISR](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L89) ...

So, I need to dig into this a bit more, but on first re-read I think there's some weirdness going on in the v6m implementation. E.g., [this load](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L116) is dead code, as the [`mrs` right after it](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L119) tramples `r0`. Without digging further, I _suspect_ that was supposed to be updating `lr` to be `MSP` so that the return at the end goes to the context switch code; [akin to what the v7m does here](https://github.com/PLSysSec/tock/blob/905ae8ced54eab523bc9365df413f8fd2bec4331/arch/cortex-v7m/src/lib.rs#L171).

> 3. How do we resume on line 406 of [the switch_to_user assembly](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L406)?

In the absence of the actual update of the `lr` noted above, I _think_ what will actually happen is the `generic_isr` will simply return to userspace, where the app will continue running until (a) its timeslice expires, or more likely (b) it makes a syscall. In either case, control will _actually_ return to the kernel along the (a) systick handler or (b) svc handler path. The scheduler [will determine there are pending interrupts](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/kernel/src/scheduler.rs#L82-L84), quit doing process things, and then the kernel will actually execute the bottom-half handler.

---

For the vast majority of applications, I believe the effect will simply be that the bottom-half interrupt handler executes slightly later than it should, but otherwise things will all still eventually execute, apps will work, etc, which is probably how this went unnoticed. (Particularly since most current userspace apps for the v6m platforms are likely to have some syscall/yield relatively quickly).

**Reply by vrindisbacher** [reply]
Thanks! Ya, this is super helpful. I didn't notice the issue with v6m as I've mostly been looking at v7m. For context, I am working on verifying these ISRs - see [here](https://github.com/vrindisbacher/tock-veri-asm-sb).

I agree that the effect is returning to the process which will continue running, and that the bottom half handler would run later than it should. I'm still interested in why we are saving registers in the v6m implementation and not the v7m though. It seems like the v7m approach is reasonable, i.e. register saving isn't necessary.

**Reply by vrindisbacher** [reply]
Also, a few follow up points here - I noticed a few things that seem interesting. 

- In the [`systick_handler` for armv6m](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L193) we put the processor in thread mode. Reading the ARM docs - _"An exception return occurs when the processor is in Handler mode and one of the following instructions loads a value of 0xFXXXXXXX into the PC"_ (See [here](https://developer.arm.com/documentation/ddi0419/c/System-Level-Architecture/System-Level-Programmers--Model/ARMv6-M-exception-model/Exception-return-behavior)). 

    The manual also explicitly mentions what happens when we load such values from thread mode: _"If an EXC_RETURN value is loaded into the PC when in Thread mode, or from the vector table, or by any other instruction, the value is treated as an address, not as a special value. This address range is defined to have XN permissions, and therefore the attempted instruction execution results in a HardFault exception."_.
    
    All of this seems to be true for `v7m` as well.  I think this is a problem. Any thoughts?

- I noticed that both in the v6m and v7m ISR's the bits from the `IPSR` are masked (see [here](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-m0/src/lib.rs#L121) and [here](https://github.com/tock/tock/blob/a1966b8ddaa1ce819b80f2f7bea466eb76e5b46c/arch/cortex-v7m/src/lib.rs#L181) respectively). This seems like it's not necessary since `ipsr` should already get us the lower bits of the `psr` register?

EDIT: 

On the first point I think I see my confusion. This doesn't set the CPU to run in thread mode, it just sets the thread mode as privileged when we return to the Kernel. 

I do think, however, that this is still an issue for the `v6m generic_isr` - we need to set Thread mode to privileged if we're coming from user space - I think this is cancelled out by the previous bug of returning to the process rather than the kernel.
